src/app/frontend/deploy/validprotocol_directive.js

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/** The name of this directive. */
+export const validProtocolValidationKey = 'validProtocol';
+
+/**
+ * Validates that chosen protocol is valid for chosen service type.
+ *
+ * @param {!angular.$resource} $resource
+ * @param {!angular.$q} $q
+ * @return {!angular.Directive}
+ * @ngInject
+ */
+export default function validProtocolDirective($resource, $q) {
+  const isExternalParam = 'isExternal';
+
+  return {
+    restrict: 'A',
+    require: 'ngModel',
+    scope: {
+      [isExternalParam]: '=',
+    },
+    link: function(scope, element, attrs, ctrl) {
+      /** @type {!angular.NgModelController} */
+      let ngModelController = ctrl;
+      let externalChanged = false;
+
+      scope.$watch(isExternalParam, (newVal, oldVal) => {
+        externalChanged = newVal !== oldVal;
```

Comment:
This statement is always true.

Btw, why do you need it? Why is the variable `externalChanged` there? I think we don't need it. It is a small short-circut optimization that gives speedup in very few cases, but leaves business logic in the validator. Adding this code means that the validator assumes that unchanged isExternal is always valid. This is true assumption today, but may not be true in the future. And I bet that whoever changes this assumption, they will not look at this code.